File: java/repository/src/main/java/validators/database/ParticipantDatabaseValidator.java
package validators.database;

import models.Participant;
import utils.JdbcUtils;
import validators.interfaces.ParticipantValidator;
import validators.interfaces.Validator;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Properties;

public class ParticipantDatabaseValidator implements ParticipantValidator {
    private JdbcUtils jdbcUtils;

    private PreparedStatement findTeamByIdPreparedStatement;
    private PreparedStatement findRallyByIdPreparedStatement;

    private static final String FIND_TEAM_BY_ID = "select * from teams where team_id=?";
    private static final String FIND_RALLY_BY_ID = "select * from rallies where rally_id=?";

    public ParticipantDatabaseValidator(Properties databaseConnectionProperties) {
        this.jdbcUtils = new JdbcUtils(databaseConnectionProperties);
    }

    @Override
    public void validate(Participant participant) {
        String errors = "";

        if (!teamOfParticipantExists(participant))
            errors += "Selected team doesn't exist!";
        if (!rallyOfParticipantIsAssignedToExists(participant))
            errors += "\nSelected rally doesn't exist!";
        if (participant.getParticipantName().equals(""))
            errors += "\nParticipant's name can't be empty!";

        if (!errors.equals(""))
            throw new Validator.InvalidModelException(errors);
    }


    private boolean teamOfParticipantExists(Participant participant) {
        try(Connection connection = jdbcUtils.getConnection();
            PreparedStatement findTeamByIdPreparedStatement = connection.prepareStatement(FIND_TEAM_BY_ID)) {
            findTeamByIdPreparedStatement.setLong(1, participant.getTeam().getId());
            try(ResultSet resultSet = findTeamByIdPreparedStatement.executeQuery()) {
              return resultSet.next();
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return false;
    }

    private boolean rallyOfParticipantIsAssignedToExists(Participant participant) {
        try(Connection connection = jdbcUtils.getConnection();
            PreparedStatement findRallyByIdPreparedStatement = connection.prepareStatement(FIND_RALLY_BY_ID)) {
            findRallyByIdPreparedStatement.setLong(1, participant.getRally().getId());
            try(ResultSet resultSet = findRallyByIdPreparedStatement.executeQuery()) {
                return resultSet.next();
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return false;
    }

}

File: csharp/Repository/Validators/Database/ParticipantDatabaseValidator.cs
using Domain.Models;
using Repository.Utils;
using Repository.Validators.Interfaces;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Repository.Validators.Database.Postgresql
{
    public class ParticipantDatabaseValidator : IParticipantValidator
    {
        private DatabaseUtils databaseUtils;
        private static readonly String findTeamByIdSqlString = "select * from teams where team_id=@team_id_param";
        private static readonly String findRallyByIdSqlString = "select * from rallies where rally_id=@rally_id_param";

        public ParticipantDatabaseValidator(IDictionary<String, String> connectionProperties)
        {
            this.databaseUtils = new DatabaseUtils(connectionProperties);
        }

        public void Validate(Participant model)
        {
            String errors = "";

            if (!TeamOfParticipantExists(model))
                errors += "Selected team doesn't exist!\n";

            if (!RallyThatParticipantIsAssignedToExists(model))
                errors += "Selected rally doesn't exist!\n";

            if (model.Name.Equals(""))
                errors += "Participant's name can't be empty!";

            if (!errors.Equals(""))
                throw new IRallyValidator.InvalidModelException(errors);
        }

        private Boolean TeamOfParticipantExists(Participant participant)
        {
            using(var connection = databaseUtils.GetConnection())
            {
                IDbCommand command = connection.CreateCommand();
                command.CommandText = findTeamByIdSqlString;
                
                IDbDataParameter teamIdParam = command.CreateParameter();
                teamIdParam.ParameterName = "@team_id_param";
                teamIdParam.Value = participant.Team.Id;

                command.Parameters.Add(teamIdParam);

                using(IDataReader reader = command.ExecuteReader())
                {
                    return reader.Read();
                }
            }
        }

        private Boolean RallyThatParticipantIsAssignedToExists(Participant participant)
        {
            using (var connection = databaseUtils.GetConnection())
            {
                IDbCommand command = connection.CreateCommand();
                command.CommandText = findRallyByIdSqlString;

                IDbDataParameter rallyIdParam = command.CreateParameter();
                rallyIdParam.ParameterName = "@rally_id_param";
                rallyIdParam.Value = participant.Rally.Id;

                command.Parameters.Add(rallyIdParam);

                using (IDataReader reader = command.ExecuteReader())
                {
                    return reader.Read();
                }
            }
        }

    }
}
